libr/bin/format/pe/pe.c

```diff
@@ -838,6 +839,69 @@ int PE_(bin_pe_get_actual_checksum)(struct PE_(r_bin_pe_obj_t)* bin) {
 	return computed_cs;
 }
 
+static char* PE_(bin_pe_get_claimed_authentihash)(struct PE_(r_bin_pe_obj_t)* bin) {
+	if (!bin->spcinfo) {
+		return NULL;
+	}
+	RASN1Binary *digest = bin->spcinfo->messageDigest.digest;
+	return r_hex_bin2strdup (digest->binary, digest->length);
+}
+
+static char* PE_(bin_pe_get_actual_authentihash)(struct PE_(r_bin_pe_obj_t)* bin) {
+	const char *hashalgo;
+	const ut8 *block, *hash = NULL;
+	char *hashstr = NULL;
+	ut32 checksum_paddr, security_entry_offset, security_dir_size;
+	ut64 len;
+	PE_(image_data_directory) * data_dir_security;
+	PE_DWord security_dir_offset;
+	RBuffer *buf;
+	RHash *ctx;
+	if (!bin->spcinfo) {
+		return NULL;
+	}
+
+	hashalgo = bin->spcinfo->messageDigest.digestAlgorithm.algorithm->string;
+	checksum_paddr = bin->nt_header_offset + 4 + sizeof(PE_(image_file_header)) + 0x40;
+	security_entry_offset =  bin->nt_header_offset + sizeof(PE_(image_nt_headers)) - 96;
+	data_dir_security = &bin->data_directory[PE_IMAGE_DIRECTORY_ENTRY_SECURITY];
+	security_dir_offset = bin_pe_rva_to_paddr (bin, data_dir_security->VirtualAddress);
+	security_dir_size = data_dir_security->Size;
+
+	buf = r_buf_new ();
+	r_buf_append_buf_slice (buf, bin->b, 0, checksum_paddr);
+	r_buf_append_buf_slice (buf, bin->b,
+		checksum_paddr + 4,
+		security_entry_offset - checksum_paddr - 4);
+	r_buf_append_buf_slice (buf, bin->b, 
+		security_entry_offset + 8,
+		security_dir_offset - security_entry_offset - 8);
+	r_buf_append_buf_slice (buf, bin->b,
+		security_dir_offset + security_dir_size, 
+		r_buf_size (bin->b) - security_dir_offset - security_dir_size);
+	block = r_buf_data (buf, &len);
+	if (!strcmp (hashalgo, "md5")) {
+		ctx = r_hash_new (true, R_HASH_MD5);
+		hash = r_hash_do_md5 (ctx, block, len);
+		hashstr = r_hex_bin2strdup (hash, R_HASH_SIZE_MD5);
+	} else if (!strcmp (hashalgo, "sha1")) {
+		ctx = r_hash_new (true, R_HASH_SHA1);
+		hash = r_hash_do_sha1 (ctx, block, len);
+		hashstr = r_hex_bin2strdup (hash, R_HASH_SIZE_SHA1);
+	}
+	r_hash_free (ctx);
+	r_buf_free (buf);
+	return hashstr;
+}
+
+char* PE_(bin_pe_get_authentihash)(struct PE_(r_bin_pe_obj_t)* bin) {
```

Comment:
maybe better to make those functions return const char *

Reply:
I am wondering when to and not to return `const char *`, because I saw some `TODO` comment suggesting to change other API to `const char *` as well, but then freeing those memory later will require re-casting them to `(char *)` which seems like code smell.

Reply:
Guess I was wrong about that, since `PE_(bin_pe_get_authentihash)` is just a public interface, and the underlying `authentihash` property is still `char *`, so freeing it won't require re-casting.